### ISSUE TITLE
Thoughts on Olympic model...

### ISSUE THREAD

**Issue by hen** [rgardler/BigLittleChallenge#6]
Random thoughts on the Olympic card model.

* Card - add a get/setProperty(String,String) API and a get/setAttribute(String,Integer) API. Child classes can build on that while using the same underlying datastore.
* Perhaps model Deck differently.  Or perhaps these are additional ideas based on not understanding what a Deck is intended to be :)
* Deck is a stack of cards, with a draw function (aka pop), plus more advanced functions like peek, push and pushToBottomOfDeck. It's ordered. Maybe there's a shuffle(). 
* Then there's Hand. A Hand is a set of cards with no order, and the player can see every card.
* Then there's Discard. This might just be a Deck that has permanent peek on it. Or there is DeckFaceDown and DeckFaceUp as children classes. 
* Also PlayedHand? A version of Hand that shows the card(s) the player is playing.
* Might need a GameBoard. Something that has N Decks etc.
* Lastly need engine style classes. Perhaps a notion of a LegalMove/RuleChecker, and the notion of play-steps. player->draw(3); player->choose_card_to_play(1) etc.

Anyway - thought I'd share the thoughts. I'm trying to think both Pokemon style, Rummy style and EuroBoard style. :)

**Comment by SorraTheOrc** [comment]
First off... patches welcome :-)

For obviously useful things we should just go ahead and issue a pull request. For things that might be a little more intrusive then an issue per topic with a proposal and subsequent discussion would be a good way to design.

RE extra functions in the deck, I agree. I'm working on this with my son and thus starting with simple hard coded attributes. The plan is to move to a flexible "attributes" model once the basic concepts are captured. He's learning to program as we go. Perhaps a good practice will be to insert "FIXME:...." notes in such places.

Agree with respect to the need to have a more complete "Deck" class. The goal is that we will later pull out the core classes (such as card and deck) into a "common" package to be used by all games. 

Currently the deck is not ordered (it's a HashMap) that's a problem, as you say it should be ordered.

In the (current) Olympics rules there is no concept of a "hand" the player can play any card from her deck. However, this will likely change in time and certainly other games will have the concept of a hand, I think this would be a good addition. There is nothing stopping us making a hand being all cards in the deck if things don't change in the Olympics rules.

Discard is a common requirement in card games. Again, the Olympics rules currently don't call for a discard deck. But that too may change in the future.

Deck, Hand, PlayedHand and Discard are all collections of cards. This seems to point to the need for a generic parent class that has the core features (collection of cards, add, remove etc.). I'm going to create an issue to break out these generic classes into a "common" module.

Olympics does not have the concept of a GameBoard (at least currently) but  quite possibly other rules would. Of course, if you have a game that needs one then go for it :-)

Agree with the need for an engine. The current code is just an initial outline, there is no actual gameplay in there yet. These will emerge. Our goal is to get to the point that we can play test the simple rules we currently have using a terminal application. For an initial playtest see https://github.com/rgardler/BigLittleChallenge/blob/master/docs/olympics/playtests.md

**Comment by SorraTheOrc** [comment]
I've moved the current code into a core module so that the basic card, deck, etc. classes can be reused in other kinds of games as discussed.

**Comment by SorraTheOrc** [comment]
Closing this issue now as none of the current rules being worked on require many of the features noted here. However, as described they are indeed useful additions. I've refactored the current code to abstract the core concepts away from the Olympics rules, this allows others to add concepts such as hands, discards etc by extding the Deck class (which should probably become the even more generic "CardCollection".

If any of these features are needed for a new set of rules lets open an issue per item and work on them. In the meantime I have implemented the properties concept on Cards as this is needed for the Olympics rules.
